proof/invariant-abstract/Invariants_AI.thy

```diff
@@ -1282,15 +1282,15 @@ definition
        (cur_sc s) s"
 
 definition valid_machine_time_2 :: "time \<Rightarrow> time \<Rightarrow> bool" where
-   "valid_machine_time_2 ct lmt \<equiv> lmt \<le> - kernelWCET_ticks - 1 \<and> ct \<le> lmt"
+   "valid_machine_time_2 ct lmt \<equiv> lmt \<le> - (kernelWCET_ticks + 3 * MAX_PERIOD) - 1 \<and> ct \<le> lmt"
```

Comment:
Ah, I see how that `3 * MAX_PERIOD` is hard-coded into the new spec for `get_cur_time`. Perhaps we could introduce an "input-only" abbreviation so that if we ever need to change this constant we can change it in one location? This is just a suggestion, I know you've done a lot of work on this already so I'm happy if this kind of optimisation is left for another time.

Reply:
The annoying this is that `getCurrentTime` will work with natural numbers, whereas `valid_machine_time` will work with words. We needed to add to the axiomatisation of `us_to_ticks` and do a fair few word proofs to get them to play nice together. 

Would it be nicer to change `getCurrentTime` to perform word addition? Seeing that the danger of wrap around would be contained to just the one spot, maybe we would be happy to do that. I'm not quite sure of the implications, but I think most of the work in AInvs about this would follow from `valid_machine_time`, which uses word addition anyway.

The other quesion is whether we're happy that these "caps" in `getCurrentTime` will always be of the form `kernelWCET_ticks + n * MAX_PERIOD` for some `n`. There's been discussion about removing the `kernelWCET_ticks` term from these things, so I don't think we'd like to go that route. Perhaps this could be further generalised to defining the cap separately. But I don't think we gain much by doing that, unless `valid_machine_time ` and `getCurrentTime` both use word addition, in which case they could both use the cap we define.

Any thoughts?

Reply:
I should have realised that all the natural number addition in `getCurrentTime` is explicity for ensuring that we return a time that is capped below the maximum 64 word. So there's no way of just performing word addition within this function.

I can make a JIRA issue with the idea of a cap that can be used in both `getCurrentTime` and `valid_machine_time`, and also mention that we may wish to have `n` as a parameter within

Reply:
https://sel4.atlassian.net/browse/VER-1394